app/models/ems_refresh/save_inventory_helper.rb

```diff
@@ -113,12 +113,28 @@ def save_child_inventory(obj, hashes, child_keys, *args)
 
   def store_ids_for_new_records(records, hashes, keys)
     keys = Array(keys)
-    hashes.each do |h|
-      r = records.detect { |r| keys.all? { |k| r.send(k) == r.class.type_for_attribute(k.to_s).cast(h[k]) } }
-      h[:id] = r.id
+    # Lets first index the hashes based on keys, so we can do O(1) lookups
+    record_index = {}
+    records.each do |record|
+      record_index[build_index_from_record(keys, record)] = record
+    end
+
+    record_class = records.first.class
```

Comment:
1. Can the class vary among records due to STI?
2. Can `type_for_attribute` vary along STI?

There are certainly STI'd tables involved (e.g. `PersistentVolume < ContainerVolume`) — but hopefully not passed to this function in the same association?  E.g. each of these calls should see uniform types:
```
    store_ids_for_new_records(ems.persistent_volumes, hashes, :ems_ref)
...elswhere...
    store_ids_for_new_records(container_group.container_volumes, hashes, :name)
```